### PR TITLE
BUG: Timestamp.unit now reflects changes in components after Timestamp.replace

### LINKED ISSUE #57749: BUG: Comparison of timestamp with same value failed
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import datetime
import pandas as pd

ts = pd.Timestamp("2023-07-15 23:08:12.134567123")
print(ts)  # shows Timestamp('2023-07-16 23:08:12.134567123')
ts2 = pd.Timestamp(datetime.datetime.utcfromtimestamp(int(ts.timestamp())))
ts2 = ts2.replace(microsecond=ts.microsecond, nanosecond=ts.nanosecond)
print(ts2)  # shows Timestamp('2023-07-16 23:08:12.134567123')
assert ts == ts2  # fails
```

### Issue Description

While repr(ts) and repr(ts2) are equal, direct comparison fails.

### Expected Behavior

ts and ts2 shall be equal.

### Installed Versions

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.9.16.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 22.5.0
Version               : Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:22 PDT 2023; root:xnu-8796.121.3~7/RELEASE_X86_64
machine               : x86_64
processor             : i386
byteorder             : little
LC_ALL                : None
LANG                  : zh_CN.UTF-8
LOCALE                : zh_CN.UTF-8

pandas                : 2.2.1
numpy                 : 1.25.2
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 61.2.0
pip                   : 22.3.1
Cython                : 3.0.2
pytest                : 8.0.2
hypothesis            : 6.46.5
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : 8.1.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : 1.3.8
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.7.2
numba                 : None
numexpr               : 2.8.4
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : 11.0.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.11.1
sqlalchemy            : 2.0.19
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : 0.21.0
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None

### PR DESCRIPTION
**Pull request by diogomsmiranda**
- [x] closes #57749
- [x] [Tests added and passed]
- [x] All [code checks passed]
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst`

The logic behind this is having the `Timestamp.unit` set as the most precise unit changed on `Timestamp.replace` if one is more precise than the previous set.

### CODE REVIEW

File: pandas/tests/scalar/timestamp/methods/test_replace.py [hunk 1]
```diff
@@ -189,3 +189,13 @@ def test_replace_preserves_fold(self, fold):
...
+
+    def test_replace_unit(self):
+        # GH#57749
+        ts = Timestamp("2023-07-15 23:08:12")
+        ts1 = Timestamp("2023-07-15 23:08:12.134567")
+        ts2 = Timestamp("2023-07-15 23:08:12.134567123")
+        ts = ts.replace(microsecond=ts1.microsecond)
+        assert ts == ts1
+        ts = ts.replace(nanosecond=ts2.nanosecond)
+        assert ts == ts2
```
Comment (jbrockmendel) [inline pandas/tests/scalar/timestamp/methods/test_replace.py:201]: can you explicitly check the result's unit

Comment (diogomsmiranda) [reply pandas/tests/scalar/timestamp/methods/test_replace.py:201]: Hey, I'll make the explicit check and keep the assertions then.

**Comment by diogomsmiranda** [comment]
Hey, I've made the requested changes.

cc @jbrockmendel

**Comment by diogomsmiranda** [comment]
@jbrockmendel @mroeschke Hey!! Sorry for the ping but it's been a while since I last heard about you, is this change still of interest?

Let me know If you need anything from me :)

**Comment by diogomsmiranda** [comment]
@jbrockmendel  I see you made a change on this files recently, does this still work as intended?
I haven't heard from you in a while. Is this still on track to be merged?

**Comment by diogomsmiranda** [comment]
Hey @mroeschke after speaking with him it is of my understanding that @jbrockmendel is taking some time mostly-off contributing for pandas, he indicated that you might be able to review this in the meantime? thanks in advance

File: doc/source/whatsnew/v3.0.0.rst [hunk 2]
```diff
@@ -361,6 +361,18 @@ Performance improvements
 
 Bug fixes
 ~~~~~~~~~
+- Fixed bug in :class:`SparseDtype` for equal comparison with na fill value. (:issue:`54770`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:364]: Looks like these all are not supposed to be here. Only your whatsnew note should exist here in the one of the sections below

Comment (diogomsmiranda) [reply doc/source/whatsnew/v3.0.0.rst]: Thanks! That was a mistake after I solved the merge conflicts. I solved it now.

File: doc/source/whatsnew/v3.0.0.rst [hunk 3]
```diff
@@ -473,6 +473,7 @@ Performance improvements
 
 Bug fixes
 ~~~~~~~~~
+- Fixed bug in :meth:`Timestamp.replace` where it would not reflect changes into :meth:`Timestamp.unit`. (:issue:`57749`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:476]: This should be in the `Datetimelike` section

**Comment by mroeschke** [comment]
Could you add a test like

```python
ts = pd.Timestamp(year=2020, month=1, day=1, nanosecond=5)
result = ts.replace(nanosecond=0)
assert result.unit == "s"
```

File: doc/source/whatsnew/v3.0.0.rst (hunk 3 above)
Comment (diogomsmiranda) [reply doc/source/whatsnew/v3.0.0.rst]: Thanks for the input!

**Comment by diogomsmiranda** [comment]
@mroeschke In fact that logic was not yet implemented, thank you for noticing! I have attempted to solve the issue by registering if a replacement with 0 was made and if so add an extra check to recalculate the replacement resolution. The test now passes successfully.

File: pandas/tests/scalar/timestamp/methods/test_replace.py [hunk 4]
```diff
@@ -192,3 +192,23 @@ def test_replace_preserves_fold(self, fold):
...
+        assert ts.unit == "us"
+        assert ts == ts1
+        ts = ts.replace(nanosecond=ts2.nanosecond)
+        assert ts.unit == "ns"
+        assert ts == ts2
+
+    def test_replace_resets_to_more_precise(self):
+        # GH#57749
+        ts = Timestamp(year=2020, month=1, day=1, nanosecond=5)
+        result = ts.replace(nanosecond=0)
```
Comment (mroeschke) [inline pandas/tests/scalar/timestamp/methods/test_replace.py:213]: Could you add test cases where a similar zeroing results in a resolution of `"ms"` and `"us"`?

**Comment by diogomsmiranda** [comment]
@mroeschke Hey, I've added the test cases requested!! Had to change some things in the logic but I believe it's working as intended now.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**